experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,389 @@
+    if (not speculativeEdges.count(edge)) {
+      toRemove.insert(edge);
+    }
+  }
+  for (CFGEdge edge : toRemove) {
+    placements.eraseCommit(edge);
+  }
+}
+
+LogicalResult PlacementFinder::findCommitPositions() {
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+
+  // We need to place a commit unit before (1) an exit unit; (2) a store
+  // unit; (3) a save unit if speculative tokens can reach them.
+  llvm::DenseSet<Operation *> visited;
+  findCommitsTraversal(visited, specPos.dstOp);
+
+  // Additionally, if there are many BBs, two control-only tokens can
+  // themselves go out of order. For this reason, additional commits need
+  // to be placed in between BBs
+  findCommitsBetweenBBs();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// SaveCommit Units Finder Methods
+//===----------------------------------------------------------------------===//
+
+// Traverse the speculator's BB from top to bottom (from the control merge until
+// the branches) and adds save-commits in such a way that every path is cut by a
+// save-commit or the speculator itself. Updates `placements`.
+void PlacementFinder::findSaveCommitsTraversal(
+    llvm::DenseSet<Operation *> &visited, Operation *currOp) {
+  // End traversal if currOp is already in visited set
+  if (auto [_, isNewOp] = visited.insert(currOp); !isNewOp)
+    return;
+
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto specBB = getLogicBB(specPos.dstOp);
+
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      if (auto succOpBB = getLogicBB(succOp); !succOpBB || succOpBB != specBB) {
+        // Stop traversal if we go outside the speculation BB
+        continue;
+      } else if (placements.containsCommit(res, succOp) or
+                 placements.containsSaveCommit(res, succOp)) {
+        // End traversal on this path, as it already is cut by a commit or sc
+        continue;
+      } else if (OpPlacement specPos = placements.getSpeculatorPlacement();
+                 specPos.srcOpResult == res and specPos.dstOp == succOp) {
+        // End traversal on this path, as it already is cut by the speculator
+        continue;
+      } else if (isa<handshake::ConditionalBranchOp>(succOp)) {
+        // A SaveCommit is needed in front of the branch
+        placements.addSaveCommit(res, succOp);
+      } else {
+        // Continue DFS traversal along the path
+        findSaveCommitsTraversal(visited, succOp);
+      }
+    }
+  }
+}
+
+LogicalResult PlacementFinder::findSaveCommitPositions() {
+  // There already exist save-commits which have been placed instead of
+  // consecutive save and commit units. Here, additional save commits are found
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto funcOp = specPos.dstOp->getParentOfType<handshake::FuncOp>();
```

Comment:
Here is one of those cases where I think `auto` is fine because the return type is also on the right-hand-side of the assignment (as a more minor reason, adding the explicit type here probably causes a line break).